utilities/transactions/lock/lock_mgr.h

```diff
@@ -0,0 +1,71 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+#ifndef ROCKSDB_LITE
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "rocksdb/utilities/transaction.h"
+#include "util/thread_local.h"
+#include "utilities/transactions/lock/lock_tracker.h"
+#include "utilities/transactions/pessimistic_transaction.h"
+
+// This file contains a generic Lock Manager interface
+
+namespace ROCKSDB_NAMESPACE {
+
+class ColumnFamilyHandle;
+
+struct DeadlockInfoBuffer {
+ private:
+  std::vector<DeadlockPath> paths_buffer_;
+  uint32_t buffer_idx_;
+  std::mutex paths_buffer_mutex_;
+  std::vector<DeadlockPath> Normalize();
+
+ public:
+  explicit DeadlockInfoBuffer(uint32_t n_latest_dlocks)
+      : paths_buffer_(n_latest_dlocks), buffer_idx_(0) {}
+  void AddNewPath(DeadlockPath path);
+  void Resize(uint32_t target_size);
+  std::vector<DeadlockPath> PrepareBuffer();
+};
+
+//
+// Base class for Point and Range-based lock manager.
+//
+class BaseLockMgr {
```

Comment:
This interface feels like for point lock only? E.g. there is no API like `TryLockRange`?

Reply:
Yes. 

Yes. The reasons for this is that class `PessimisticTransactionDB` has several places where it "knows" which kind of lock manager it is using and so it calls to `RangeLockMgr` : 

```c++
void PessimisticTransactionDB::init_lock_manager() {
    range_lock_mgr_ = lock_mgr_->getRangeLockMgr();
```

```c++
Status PessimisticTransactionDB::TryRangeLock(PessimisticTransaction* txn,
                                              uint32_t cfh_id,
                                              const Endpoint& start_endp,
                                              const Endpoint& end_endp) {
  if (range_lock_mgr_) {
    return range_lock_mgr_->TryRangeLock(txn, cfh_id, start_endp, end_endp,
                                         /*exclusive=*/true);
  } else
    return Status::NotSupported();
}
```

```c++
void PessimisticTransactionDB::UnLock(PessimisticTransaction* txn,
                                      const LockTracker& keys,
                                      bool all_keys_hint) {
  if (all_keys_hint && range_lock_mgr_) {
    range_lock_mgr_->UnLockAll(txn, GetEnv());
    return;
  }
  lock_mgr_->UnLock(txn, keys, GetEnv());
}
```

Should this code be changed so that the calls are made through the `BaseLockMgr`  pointer and all difference is hidden in the lock manager implementation?